DEC-381: Responsive flex classes with responsive gutters

sherakama opened the pull request:
# READY FOR REVIEW

# Summary
- Improves the developer experience in using the flex based classes so that the developer does not have to add the size for every breakpoint
- Breakpoint based flex size now has responsive behavior for all sizes above the one in use.

# Needed By (Date)
- End of the week

# Urgency
- Medium

# Steps to Test

1. Review code
2. Compile sass
3. Compile styleguide
4. Review core flex-box example for consistency with the existing version
5. Review flex-box layouts for issues

Test new behavior:
1. Create a new flexbox container and child element
2. Add different flexbox classes and review for expected behavior

# Affected Projects or Products
- Flex classes
- Flex layouts

# Associated Issues and/or People
- #381 

# See Also
- [PR Checklist](https://gist.github.com/sherakama/0ba17601381e3adbe0cad566ad4d80a5)

yvonnetangsu commented on core/src/templates/layout/flex/three-column/battleship/battleship.twig:
```diff
@@ -24,49 +24,49 @@
 
     {% if second is not empty %}
-    <div class="flex-md-6-of-12 flex-lg-4-of-12 flex-xl-4-of-12 flex-2xl-4-of-12">
+    <div class="flex-md-6-of-12 flex-lg-4-of-12">
```
Haven't taken a really close look but Battleship has some issues - something to do with it's md-6 but lg-4?

sherakama replied:
Oh, this might be because the 4-of-12 is rendered before the 6-of-12 items as the loop goes through grid first then breakpoint. I'll refactor to see if that helps.

yvonnetangsu commented on core/src/templates/layout/flex/four-column/molive/molive.twig:
```diff
@@ -30,19 +30,19 @@
       </div>
 
-      <div class="flex-md-6-of-12 flex-lg-6-of-12 flex-xl-3-of-12 flex-2xl-3-of-12">
+      <div class="flex-md-6-of-12flex-xl-3-of-12">
```
```suggestion
      <div class="flex-md-6-of-12 flex-xl-3-of-12">
```

sherakama commented:
@yvonnetangsu ready for another pass.

yvonnetangsu commented:
Found one thing about flex grid that's on the live site also, so it's not because of your work in this branch. Under Core - > flex grid, the cells that say x-of-12-for-all seem to behave the same as x-of-12-for-sm-and-up, ie, they don't really do x-of-12 at xs. Is this intentional?

![Screen Shot 2019-06-11 at 11 23 11 AM](https://user-images.githubusercontent.com/42749717/59296757-9444b280-8c3b-11e9-95e3-12ad5280e655.png)

yvonnetangsu commented:
For the push classes, looks like .flex-push-xs-1, .flex-push-sm-1, .flex-push-md-1 etc all do the same thing (set the same margin-left percentage). Could we just have .flex-push-1 etc (get rid of the breakpoint specific classes)?
![Screen Shot 2019-06-11 at 11 53 15 AM](https://user-images.githubusercontent.com/42749717/59298627-c7894080-8c3f-11e9-8b0a-ebdd2ea9971d.png)

sherakama commented:
> I can see that this makes sense (at XS, everything takes up 100% and just stack on top of each other). In that case, perhaps we could eliminate the .flex-sm-x-of-12 classes since they're the same as the .flex-x-of-12 ones?

Perhaps we can now, prior to this branch, if you had a `flex-6-of-12` and a `flex-sm-3-of-12` you would get 3-of-12 only on the small and 6-of-12 everywhere else. Now, with the growing cascade, they pretty much do the same.

sherakama commented:
> For the push classes, looks like .flex-push-xs-1, .flex-push-sm-1, .flex-push-md-1 etc all do the same thing (set the same margin-left percentage). Could we just have .flex-push-1 etc (get rid of the breakpoint specific classes)?

No, I don't think so, as they activate at different breakpoints. eg: push at md but not at lg.

yvonnetangsu commented:
> > For the push classes, looks like .flex-push-xs-1, .flex-push-sm-1, .flex-push-md-1 etc all do the same thing (set the same margin-left percentage). Could we just have .flex-push-1 etc (get rid of the breakpoint specific classes)?
> 
> No, I don't think so, as they activate at different breakpoints. eg: push at md but not at lg.

Good point!

sherakama commented:
@yvonnetangsu I removed the `flex-sm-x-of-y` classes and refactored the collapse and gap code to not assume our breakpoint codes as in the future we might want to add an `xxs` or something.

yvonnetangsu commented:
> @yvonnetangsu I removed the `flex-xs-x-of-y` classes and refactored the collapse and gap code to not assume our breakpoint codes as in the future we might want to add an `xxs` or something.

Thanks for the change, @sherakama ! This looks good. I do think I need a bit more time to test this out more thoroughly. I'll look at it some more tonight.

yvonnetangsu commented on core/dist/css/decanter-grid.css:
```diff
@@ -674,498 +674,1590 @@ body {
-@media only screen and (min-width: 0) and (max-width: 575px) {
-  .flex-xs-12-of-12 {
+@media only screen and (min-width: 576px) {
+  .flex-sm-12-of-12 {
```
Is it possible to do this optimization - for all the `.flex-xx-12-of-12` classes, only write the first media query `@media only screen and (min-width: 576px)` since all the subsequent media queries generate the same numbers?
![Screen Shot 2019-06-12 at 1 55 17 AM](https://user-images.githubusercontent.com/42749717/59337922-4a4ce280-8cb6-11e9-8e4b-0224e9eccf89.png)

yvonnetangsu commented on core/dist/css/decanter-grid.css:
```diff
@@ -674,498 +674,1590 @@ body {
+            flex: 0 0 calc(8.3333333333% - 44px);
+    max-width: calc(8.3333333333% - 44px); } }
+
+@media only screen and (min-width: 576px) {
```
It looks like for example, the `.flex-xl-x-of-12` classes start with media query for the `xl` breakpoint (which is right), but the `.flex-2xl-x-of-12` classes start with media query for the `xs` breakpoint? Seems like they should just start at `2xl` breakpoint?
![Screen Shot 2019-06-12 at 1 59 49 AM](https://user-images.githubusercontent.com/42749717/59338345-0a3a2f80-8cb7-11e9-99e6-a9857da218f4.png)

yvonnetangsu replied:
Yup, something's weird with the `2xl `and up classes 👀  Screenshot is at `sm` breakpoint.
![Screen Shot 2019-06-12 at 2 14 53 AM](https://user-images.githubusercontent.com/42749717/59338879-0b1f9100-8cb8-11e9-9758-63292bdb7390.png)

yvonnetangsu replied:
This one works great - thanks!

yvonnetangsu commented:
Got a question about the push classes, @sherakama. Posting here since you're concentrating 😃 
Am I doing something wrong here? It looks like the push value is a bit off? Also, do we need the xs push classes since everything is 100% width?

![Screen Shot 2019-06-12 at 10 18 51 AM](https://user-images.githubusercontent.com/42749717/59372159-bc93e600-8cfb-11e9-83df-ea97af90350f.png)

sherakama commented:
> Got a question about the push classes, @sherakama. Posting here since you're concentrating 😃
> Am I doing something wrong here? It looks like the push value is a bit off? Also, do we need the xs push classes since everything is 100% width?

Good questions. I will review.
